### ISSUE TITLE
Evaluate differences between parsers (Acorn, Babylon, Espree)

### ISSUE THREAD

**Issue by mnkhouri** [depcheck/depcheck#269]
The project is currently using the [Babylon](https://github.com/babel/babylon/blob/6.x/README.md) parser. This choice seems to have been made due to [shortcomings in Acorn](https://github.com/lijunle/depcheck-es6/commit/9b26e19d8f5203ae55bdb1fc8770cf9ea42f00ac) as of 2015.

It would be interesting to revisit the differences between these three parsers, to see:
- how coupled is the project to the output of the Babylon parser?
- is there a major performance difference?
- is there a major difference in supported options / plugins ([Babylon](https://babeljs.io/docs/en/next/babel-parser.html), [Acorn](https://github.com/acornjs/acorn#main-parser))?

This issue is prompted by #259, which could be resolved by having an ES5 parser separate from an ES6 parser (which Acorn would allow us to do using its [ecmaVersion option](https://github.com/acornjs/acorn#main-parser))

**Comment by LinusU** [comment]
One thing that I've had problems with acorn in the past is that they are very strict on pulling in features that aren't yet stabilized:

> NOTE: Only 'stage 4' (finalized) ECMAScript features are being implemented by Acorn.

e.g. object spread would make it choke on the entire file for a long time, even though most people used that syntax...

**Comment by rumpl** [comment]
What about [espree](https://github.com/eslint/espree)?

**Comment by LinusU** [comment]
Hmm, actually I think that this is working in Babylon (now called `@babel/parser` - https://babeljs.io/docs/en/next/babel-parser.html).

`async` (and `await`) are reserved keywords only when the source type is "module", setting that to "script" makes it work. This makes sense as this is in line with the ECMAScript spec.

Unfortunately, detecting whether a file is `script` or `module` isn't as easy as it would seem...

To be honest, I don't think that the issue raised is super important to fix. It would be wise to migrate the code to stop using `async` and `await` as variable names anyhow...

**Comment by mnkhouri** [comment]
> What about espree?

I'll edit the issue to include `espree`, probably good to include it in the comparison

> To be honest, I don't think that the issue raised is super important to fix. It would be wise to migrate the code to stop using async and await as variable names anyhow...

I agree the linked issue isn't super important, but I wonder if we will run into other issues down the road related to backwards incompatability in ES specs. Trivial example:

```
const acorn = require('acorn')
const programText = 'var class = 42;'
let ast = acorn.parse(programText, {
    ecmaVersion: 3  // no error
});
ast = acorn.parse(programText, {
    ecmaVersion: 8 // SyntaxError: Unexpected keyword 'class'
});
```

Right now, we have an [`es6` and `es7` parser](https://github.com/depcheck/depcheck/tree/master/src/parser), which is implemented as Babylon with different plugins, but using Acorn (or maybe espree?) would allow us to explicitly have a parser for each different ES version (since Acorn supports that `ecmaVersion` option.

TL;DR: are there often backwards-incompatible ES spec differences? If "yes", it's probably worth having a parser that can explicitly target an older spec. However, I think the answer is "no"
